src/Npgsql/Internal/PgReader.cs

```diff
@@ -0,0 +1,720 @@
+        buffer = _charsReadBuffer;
+
+        return false;
+    }
+
+    internal void ResetCharsRead(out int charsRead)
+    {
+        if (!IsCharsRead)
+            throw new InvalidOperationException("No active chars read");
+
+        switch (_charsReadReader)
+        {
+            case PreparedTextReader reader:
+                reader.Restart();
+                break;
+            case StreamReader reader:
+                reader.BaseStream.Seek(0, SeekOrigin.Begin);
+                reader.DiscardBufferedData();
+                break;
+        }
+        _charsRead = charsRead = 0;
+    }
+
+    internal void AdvanceCharsRead(int charsRead)
+    {
+        _charsRead += charsRead;
+        _charsReadOffset = null;
+        _charsReadBuffer = null;
+    }
+
+    internal void InitCharsRead(int dataOffset, ArraySegment<char>? buffer, out int? charsRead)
+    {
+        if (!Resumable)
+            throw new InvalidOperationException("Wasn't initialized as resumed");
+
+        charsRead = _charsReadReader is null ? null : _charsRead;
+        _charsReadOffset = dataOffset;
+        _charsReadBuffer = buffer;
+    }
+
+    internal PgReader Init(int fieldLength, DataFormat format, bool resumable = false)
+    {
+        if (resumable)
+        {
+            if (Resumable)
+            {
+                Debug.Assert(Initialized);
+                return this;
+            }
+            _resumable = true;
+        }
+        else if (Initialized)
+            ThrowHelper.ThrowInvalidOperationException("Cannot be initialized to be non-resumable until a commit is issued.");
+
+        Debug.Assert(!Initialized || (Resumable && resumable), "Reader wasn't properly committed before next init");
+        Debug.Assert(!_requiresCleanup, "Should not be possible to reach");
+
+        _readStarted = false;
+        _fieldStartPos = _buffer.CumulativeReadPosition;
+        _fieldFormat = format;
+        _fieldSize = fieldLength;
+        return this;
+    }
+
+    internal void StartRead(Size bufferRequirement)
+    {
+        Debug.Assert(FieldSize >= 0);
+        _readStarted = true;
+        _fieldBufferRequirement = bufferRequirement;
+        if (bufferRequirement.GetValueOrDefault() is not 0)
+            Buffer(bufferRequirement);
+    }
+
+    internal ValueTask StartReadAsync(Size bufferRequirement, CancellationToken cancellationToken)
+    {
+        Debug.Assert(FieldSize >= 0);
+        _readStarted = true;
+        _fieldBufferRequirement = bufferRequirement;
+        return bufferRequirement.GetValueOrDefault() is not 0 ? BufferAsync(bufferRequirement, cancellationToken) : new();
+    }
+
+    internal void EndRead()
+    {
+        if (_resumable)
+            return;
+
+        // If it was upper bound we should consume.
+        if (_fieldBufferRequirement is { Kind: SizeKind.UpperBound })
+        {
+            Consume(FieldRemaining);
+            return;
+        }
+
+        if (FieldOffset < FieldSize)
```

Comment:
```suggestion
        if (FieldOffset != FieldSize)
```

Reply:
It would be a confusing error if we throw not everything was consumed for an overconsumption. Additionally, during debug that is, we can't actually consume more as we do bounds checks.

Reply:
I've tuned the exception a bit and changed the check